packages/@sanity/form-builder/src/inputs/BlockEditor-slate/util/prepareSlateForBlockEditor.js

```diff
@@ -92,55 +101,145 @@ export default function prepareSlateForBlockEditor(blockEditor) {
+              )
+            ) ? document : null
+        },
+        normalize: (change, document) => {
+          change.deselect()
+          const hasEmptySingleContentBlock = document.nodes.size === 1
+          change.insertNodeByKey(document.key, 0, Block.create(SLATE_DEFAULT_BLOCK))
+          if (hasEmptySingleContentBlock) {
+            change.removeNodeByKey(document.nodes.first().key)
+          }
+          return change.collapseToStartOf(change.state.document.nodes.first()).focus()
+        }
+      },
+      // Rule to ensure that every non-void block has a style
+      {
+        match: node => {
+          // contentBlock with no style
+          if (node.kind === 'block' && !node.isVoid && node.text && node.data.get('style') === undefined
+          ) {
+            return node
+          }
+          return undefined
+        },
+        validate: contentBlock => {
+          return contentBlock
+        },
+        normalize: (change, contentBlock) => {
+          const data = {...contentBlock.data.toObject(), style: BLOCK_DEFAULT_STYLE}
+          return change
+            .setNodeByKey(contentBlock.key, {type: 'contentBlock', ...{data}})
+        }
+      },
+      // Rule to ensure that annotation _key's within a block is unique
+      // Duplication can happen when copy/pasting annotation spans within the same block
+      {
+        match: node => {
+          // contentBlock with annotations
+          return node.kind === 'block'
+            && node.type === 'contentBlock'
+            && node.filterDescendants(desc => {
+              const annotations = desc.data && desc.data.get('annotations')
+              return annotations && Object.keys(annotations).length
+            }).size
+        },
+
+        validate: contentBlock => {
+          // return the last occurence of nodes with annotations that has the same _key
+          const duplicateKeyNodes = contentBlock.filterDescendants(
+            desc => desc.data && desc.data.get('annotations')
+          )
+          .toArray()
+          .map(aNode => {
+            const annotations = aNode.data.get('annotations')
+            return Object.keys(annotations).map(name => annotations[name]._key)
+          })
+          .reduce((a, b) => {
+            return a.concat(b)
+          }, [])
+          .filter((key, i, keys) => keys.lastIndexOf(key) !== i)
+          if (duplicateKeyNodes.length) {
+            return duplicateKeyNodes.map(key => {
+              return {
+                dupKey: key,
+                dupNode: contentBlock.filterDescendants(
+                  desc => {
+                    const annotations = desc.data && desc.data.get('annotations')
+                    return annotations
+                      && Object.keys(annotations)
+                          .find(name => annotations[name]._key === key)
+                  }).toArray().slice(-1)[0] // Last occurence
```

Comment:
since this is an immutable collection, could you just call `.last()` instead of `toArray().slice(...)` here?